### PR TITLE
Promise support

### PR DESCRIPTION
**Pull request by MatthieuLemoine**
This PR adds Promise support as mentioned in #53.

## Promise support 
- getUserMedia
- MediaStreamTrack.getSources
- createOffer
- createAnswer
- setLocalDescription
- setRemoteDescription
- addIceCandidate
- getStats

## Javascript
Using @philikon's idea, I defined a ``withLegacyCallbacks`` function to allow the user to choose between the callbacks and the promise version.

## Native
All the corresponding native methods now returns promises.

## Documentation
The README has been updated with examples using promises instead of callbacks.

### CODE REVIEW

**Comment by saghul** [comment]
Hey, great work, thanks for contributing! I'll review as soon as I can squeeze in some time! (trying to wrap up a the M57 release since M58 was just announced).

**Comment by notedit** [comment]
M58 released,  any update for this.

**Comment by MatthieuLemoine** [comment]
Waiting for a review from @saghul

**Comment by notedit** [comment]
ping @saghul    what else we need to merge this?

**Comment by saghul** [comment]
Ouch, sorry folks, it fell through the cracks, I'll make sure to review it by next week. 🙏

**Comment by minhtc** [comment]
awesome

**Comment by notedit** [comment]
thanks @saghul .   sorry to ping you.   :)

**Comment by ugiacoman** [comment]
Any update?

**Comment by saghul** [comment]
Not really. But I'm afraid master changed since, and the PR no longer applies cleanly :-S

**Comment by MatthieuLemoine** [comment]
@saghul Conflicts fixed

**Comment by saghul** [comment]
@MatthieuLemoine thanks a lot!

**Comment by notedit** [comment]
is  anyone  reviewing this ?

**Comment by zxcpoiu** [comment]
@notedit 
I think @saghul has a plan to review it.

I will find time to review it as well.

File: getUserMedia.js [hunk 1]
```diff
@@ -38,39 +30,31 @@ export default function getUserMedia(
...
       for (const track of tracks) {
         stream.addTrack(new MediaStreamTrack(track));
       }
-
-      successCallback(stream);
-    },
-    /* errorCallback */ (type, message) => {
+      return stream;
+    })
+    .catch(({ message, code }) => {
```
Comment (zxcpoiu) [inline getUserMedia.js]: shouldn't it be `{ code, message }` ?

**Comment by MatthieuLemoine** [comment]
@zxcpoiu It should not include any breaking changes. I've done it with backward compatibility in mind. I've been testing it since April but it may be good if others try it before merging as I'm not using all the react-native-webrtc's features.

(For the emoji, credits to [gitmoji](https://gitmoji.carloscuesta.me/))

**Comment by chroth7** [comment]
FWIW, I just used this branch in a RN app we are currently developing, and it works very well.

As we are in parallel doing a web-version (using promises), we had no problems porting the code to RN, literally in minutes, based on this branch.

The only change we had to do (which is unrelated to this PR!) is to use `addStream` (deprecated: https://developer.mozilla.org/de/docs/Web/API/RTCPeerConnection/addStream)

```javascript
// React-Native version (this lib)
peerConnection.addStream(stream);
```

```javascript
// Web-Version
const audioTracks = stream.getAudioTracks();
peerConnection.addTrack(audioTracks[0], stream);
```

I might open an issue for that

Sorry for spamming. But good work here!

**Comment by zxcpoiu** [comment]
@chroth7 

thanks a lot for your feedback. 
I would merge it at the end of this week if there is no other feedback. :+1:

File: getUserMedia.js (hunk 1 above)
Comment (chroth7) [reply getUserMedia.js]: That shouldn't matter:

```javascript
const obj = {a: 1, b: 2};
const f1 = ({a, b}) => console.log(a);
const f2 = ({b, a}) => console.log(a);
f1(obj); // => 1
f2(obj); // => 1
```

File: RTCPeerConnection.js [hunk 2]
```diff
@@ -195,24 +195,31 @@ export default class RTCPeerConnection extends EventTarget(PEER_CONNECTION_EVENT
...
-      // It turns out that on Android it is faster to construct a single
-      // JSON string representing the array of StatsReports and have it
-      // pass through the React Native bridge rather than the array of
-      // StatsReports.
-      if (typeof stats === 'string') {
-        try {
-          stats = JSON.parse(stats);
-        } catch (e) {
-          throw e;
+  getStats(track, success, failure) {
```
Comment (zxcpoiu) [inline RTCPeerConnection.js]: @MatthieuLemoine 

my bad, should be `getStats(track) {`
help~

**Comment by zxcpoiu** [comment]
@MatthieuLemoine

I have it rebased to master. please kindly double check if I'm doing wrong.

oh, I missed one param, please help me to correct it. :(

File: getUserMedia.js (hunk 1 above)
Comment (zxcpoiu) [reply getUserMedia.js]: @chroth7 

your right, my eyes was ignoring `{}` mysteriously
thanks

**Comment by MatthieuLemoine** [comment]
@zxcpoiu I fixed it

**Comment by zxcpoiu** [comment]
landed.

leave it on master for a while and release a new version then.

**Comment by zxcpoiu** [comment]
big thanks to :sparkles: @MatthieuLemoine :sparkles:

**Pull request merged**